src/main/java/net/glowstone/net/handler/play/player/PlayerAbilitiesHandler.java

```diff
@@ -15,7 +18,20 @@ public void handle(GlowSession session, PlayerAbilitiesMessage message) {
 
         GlowPlayer player = session.getPlayer();
         boolean flying = (message.getFlags() & 0x02) != 0;
-
-        player.setFlying(player.getAllowFlight() && flying);
+        boolean isFlying = player.isFlying();
+        boolean canFly = player.getAllowFlight();
+        if (isFlying != flying) {
+            if (!flying || canFly) {
```

Comment:
Could you describe the logic in a comment? i.e. Fire the event if either the player disables flight, or enables it and is allowed to do so.